addons/folo/jaxrs/src/main/java/org/commonjava/indy/folo/bind/jaxrs/FoloNPMContentAccessResource.java

```diff
@@ -0,0 +1,266 @@
+    {
+        final TrackingKey tk = new TrackingKey( id );
+        final String baseUri = uriInfo.getBaseUriBuilder().path( BASE_PATH ).path( id ).build().toString();
+
+        EventMetadata metadata = new EventMetadata().set( TRACKING_KEY, tk )
+                                                    .set( ACCESS_CHANNEL, AccessChannel.NPM_REPO )
+                                                    .set( STORAGE_PATH,
+                                                          Paths.get( packageName, PACKAGE_JSON ).toString() );
+
+        MDC.put( CONTENT_TRACKING_ID, id );
+
+        return handler.doGet( NPM_PKG_KEY, type, name, packageName, baseUri, request, metadata );
+    }
+
+    @ApiOperation( "Retrieve and track NPM file/artifact content under the given artifact store (type/name) and path." )
+    @ApiResponses( { @ApiResponse( code = 404, message = "Content is not available" ),
+                           @ApiResponse( code = 200, response = String.class,
+                                         message = "Rendered content listing (when path ends with '/index.html' or '/')" ),
+                           @ApiResponse( code = 200, response = StreamingOutput.class, message = "Content stream" ), } )
+    @GET
+    @Path( "/{packageName}/{versionTarball: (.*)}" )
+    public Response doGet( @ApiParam( "User-assigned tracking session key" ) @PathParam( "id" ) final String id,
+                           @ApiParam( allowableValues = "hosted,group,remote", required = true ) @PathParam( "type" )
+                           final String type, @PathParam( "name" ) final String name,
+                           final @PathParam( "packageName" ) String packageName,
+                           final @PathParam( "versionTarball" ) String versionTarball,
+                           @Context final HttpServletRequest request, @Context final UriInfo uriInfo )
+    {
+        final TrackingKey tk = new TrackingKey( id );
+        EventMetadata metadata =
+                new EventMetadata().set( TRACKING_KEY, tk ).set( ACCESS_CHANNEL, AccessChannel.NPM_REPO );
+        MDC.put( CONTENT_TRACKING_ID, id );
+
+        final String path = Paths.get( packageName, versionTarball ).toString();
+        final String baseUri = uriInfo.getBaseUriBuilder().path( BASE_PATH ).path( id ).build().toString();
+
+        return handler.doGet( NPM_PKG_KEY, type, name, path, baseUri, request, metadata );
+    }
+
+    @ApiOperation( "Retrieve and track NPM root listing under the given artifact store (type/name)" )
+    @ApiResponses( { @ApiResponse( code = 404, message = "Content is not available" ),
+                           @ApiResponse( code = 200, response = String.class,
+                                         message = "Rendered root content listing" ),
+                           @ApiResponse( code = 200, response = StreamingOutput.class, message = "Content stream" ), } )
+    @GET
+    @Path( "/" )
```

Comment:
Do we need to support root listings via the tracked API? With the redirect to the browse API on the Maven side, I'm not sure if a listing request would actually get tracked. I'm not sure what the value of supporting this is, TBH.

Reply:
I'm not sure if there is case that will do listing on root path. I agree to remove this. Will do.